library implementation with tests, doc and build

ntarocco opened the pull request:
For reviewers, a few notes:
* We had a few discussions about how to use `Overridable` and one was that maybe it could be handy to have a "store", singleton object, that contains the overridden components.
In Invenio, we want to keep user customizations as easy as possible, even for users with no JS experience. With the global store, user can create a random JS file (from a template) and add cmps to the store, that's it. In the Jinja template then we will have:
```
<script type="override.js" /> // if no overrides, this file will not exist
<script type="app.js" />
```
This will avoid to have to change the `app.js`, no need to know if it has to import or not some `map` of overridden components. The drawback I see with this is that it might happen that, by mistake, you add components to the store *after* the component is rendered, something to take into account.
So... I am not very happy with the solution, appreciate your input!
* I changed for React Context, please check if this is what you meant, have a look to the tests.
* Tests only have the `parametrized` now because I would like to have your validation before completing them!

ThiefMaster commented on rollup.config.js:
```diff
@@ -0,0 +1,41 @@
+  ],
+  plugins: [
+    peerDepsExternal(),
+    postcss({
+      plugins: [],
+      minimize: true,
+      sourceMap: "inline",
+    }),
```
probably not needed since no css is involved here

ntarocco replied:
Yes, sorry, I will clean up.

ThiefMaster commented on src/index.js:
```diff
@@ -1,58 +1,74 @@
+    this.components[id] = Component;
+  };
+
+  get = (id) => {
+    return this.components[id];
+  };
```
What about making the context dict-like so you can use `overrideStore[id]` to access it? This would allow using a normal dict when using a custom value for the context - especially since the store class is not exported.

If you don't intend to support this, then I'd get rid of the context and use the store directly.

ntarocco replied:
See answer below.

ThiefMaster commented on src/index.js:
```diff
@@ -1,58 +1,74 @@
+    this.components = {};
+  };
+}
+export const overrideStore = new OverriddenComponentsStore();
+export const OverridableContext = React.createContext(overrideStore);
```
The default passed to `createContext` is only used when no provider for the context is present - so I think we need to decide which behavior we want in this case:

- do not override anything if the context  is not provided
- always apply defined overrides, regardless of whether or not the context is provided

If we go for the first option, I wouldn't pass a default here.

If we go for the second option, then I propose removing the context altogether **or** being clear that the context provider should usually *not* be specified, unless you want to use a custom store for your overrides instead of the global one.

ntarocco replied:
Agree with you, I would remove the default.
The main question is about your previous comment and this is exactly what I was trying to explain in the notes.
When I took the `Overridable` from Indico and adapted to our needs, I removed hooks+redux and replaced with a global store everywhere, no context. Personally, I am not very happy with this global store and I find Context nicer, but it solves our problem.

With global store you can do from anywhere, any js file this:
```
import {store} ...
store.add(id, MyCustomCmp)
```
and then in our internal `app.js` we don't have to do anything because `Overridable` will use the global store.

With React Context, you will have to pass the map to the `<Context.Provider value={map}`. Very easy in a SPA, but in Invenio is harder: when building JS assets, the internal `app.js` does not know if the user has created an extra `.js` to override components. Basically I cannot have in `app.js` (like in Python for example if you allow me the pseudocode to explain what I mean):
```
try {
  import overrides from './overrides.js';
} catch {
   // no overrides
}
...
<Context.Provider value={overrides}...
```
I would have to inject this extra obj `overrides` via HTML DOM... Is it maybe more clear? Let me know if you have a better solution in mind...

ThiefMaster replied:
I think you could always have the overrides locally in your application and just import it from both the place where you pass it to the <Context.Provider>` and the places where you want to add new overrides.

But I do see the benefit of having a global list of overrides. ;)

ThiefMaster replied:
But at least when you use plugins there's also an argument against the global override list: If you define one entry point for all plugins and let them "register" their overrides, you really don't want to have to deal with shared state - especially since it may not work well when the plugins and core are built as separate webpack projects.

ntarocco commented:
I pushed the changes. It now contains the overridable lib and the global store as utils.
Added tests and doc.
Can you please review:
* code
* tests
* doc/README
* build via Rollup

**Be aware** that it does not contain the `withRouter` as in Indico.

mvidalgarcia commented on README.md:
````diff
@@ -1 +1,110 @@
+
+To run tests:
+```
+npm run tests
````
```suggestion
npm run test
```

mvidalgarcia commented on src/store.js:
```diff
@@ -0,0 +1,31 @@
+/**
+ * Simple utility class responsible of keeping track of all overridden components.
+ * @constructor object containing the initial map `id: Component` of overriden components
```
```suggestion
 * @constructor object containing the initial map `id: Component` of overridden components
```

mvidalgarcia commented on src/overridable.test.js:
```diff
@@ -0,0 +1,208 @@
+      title: 'New title',
+      color: 'red',
+    });
+    const overriddenCmps = {ExampleComponent: parametrized};
```
I know that it works like this, but to be consistent with the example on README, what about using an explicit string here?
`const overriddenCmps = {'ExampleComponent': parametrized};`

ntarocco replied:
It is ESLint + Prettier that is removing it :)

ThiefMaster replied:
I'd rather remove the quotes from the docs ;)

mvidalgarcia replied:
Oh, then we can do the other way round, modify [this](https://github.com/indico/react-overridable/pull/1/files#diff-04c6e90faac2675aa89e2176d2eec7d8R58)

ntarocco replied:
Changed in the doc!

mvidalgarcia commented on src/store.test.js:
```diff
@@ -0,0 +1,55 @@
+    const mounted = mount(
+      <OverridableContext.Provider value={overrideStore.getAll()}>
+        <OverridableExampleComponent title="Ignored title" color="red">
+          <ul>
+            <li>It should NOT appear!</li>
+          </ul>
```
Are you missing `children` in `ExampleComponent`? Otherwise I think this will never be taken into account. It could also be completely removed.

ntarocco replied:
Thanks, fixed!

lnielsen commented:
Looks good to me (though not an expert ;-)

ntarocco commented on README.md:
```diff
@@ -1 +1,110 @@
+
+With `react-overridable` you can mark your React components as overridable
+and allow other apps to customize them. This can be useful when creating
+libraries with a default components implementation which require to be
```
Here I wanted to say "a default implementation of components". Should I re-phrase it instead?

mvidalgarcia replied:
Yes, that sounds better.

ntarocco replied:
Fixed!

ThiefMaster commented on .eslintrc.yml:
```diff
@@ -0,0 +1,38 @@
+  import/no-unresolved: off
+  jest/expect-expect:
+    - warn
+    - assertFunctionNames:
+      - expect
+      - expectURL
+      - expectURLError
+      - expectURLConv
+      - expectURLConvError
+      - expectTranspiledPlugin
+      - expectTranspiledMacro
+      - expectTranspiledPluginMocked
+      - expectTranspiledMacroMocked
```
not needed (except *maybe* `expect` itself)

ThiefMaster commented on README.md:
```diff
@@ -1 +1,110 @@
+import React, {Component} from 'react';
+import Overridable, {parametrize, OverridableContext} from 'react-overridable';
+
+class TitleComponent extends Component {
```
Eventually I'd like the readme to use a function component (and maybe an example with a class component somewhere else). Not super urgent though.

ntarocco replied:
I think there is really room for improvements and I can spend sometime later on to make it much better in general if you agree.

ThiefMaster commented on README.md:
````diff
@@ -1 +1,110 @@
+const overriddenCmps = {TitleComponent: parametrized};
+```
+
+2. **Provide new render elements**: override the default rendered elements for the marked sections. Props are passed and can be used in the new template.
````
some of the lines in here are pretty long

ThiefMaster commented on src/overridable.js:
```diff
@@ -0,0 +1,86 @@
+  const overriddenCmps = useContext(OverridableContext);
+  const Overridden = overriddenCmps[id];
+
+  return Overridden !== undefined
+    ? // If there's an override, we replace the component's content with
+      // the override + props
+      React.createElement(Overridden, {...childProps, ...restProps})
+    : // No override? Clone the Overridable component's original children
+    child
+    ? React.cloneElement(child, childProps)
+    : null;
+}
```
What about making this a bit more readable? 

```js
if (Overridden !== undefined) {
  // If there's an override, we replace the component's content with
  // the override + props
  return React.createElement(Overridden, {...childProps, ...restProps});
} else if (child) {
  // No override? Clone the Overridable component's original children
  return React.cloneElement(child, childProps);
} else {
  return null;
}
```

ntarocco replied:
LOL I rewrote before reading your comment and they look almost the same, good ;)

ThiefMaster commented on src/overridable.js:
```diff
@@ -0,0 +1,86 @@
+ * React component to enable overriding children when rendering.
+ */
+function Overridable({id, children, ...restProps}) {
+  const child = React.Children.only(children); // throws an error if multiple children
```
This breaks in case of no children. The old code had this to avoid it:

    const child = children ? React.Children.only(children) : null;

ntarocco replied:
fixed!

ThiefMaster commented on src/overridable.js:
```diff
@@ -0,0 +1,86 @@
+  const name = Component.displayName || Component.name;
+  ParametrizedComponent.displayName = `Parametrized(${name})`;
+  return ParametrizedComponent;
+}
```
let's put an empty line between functions

ThiefMaster commented on src/overridable.js:
```diff
@@ -0,0 +1,86 @@
+ * React component to enable overriding children when rendering.
+ */
+function Overridable({id, children, ...restProps}) {
+  const child = React.Children.only(children); // throws an error if multiple children
+  const childProps = child.props || {};
+  const overriddenCmps = useContext(OverridableContext);
+  const Overridden = overriddenCmps[id];
```
this should fix it (also moved up the hook call to keep them at the beginning)

```suggestion
  const overriddenCmps = useContext(OverridableContext);
  const child = children ? React.Children.only(children) : null;
  const childProps = child ? child.props : {};
  const Overridden = overriddenCmps[id];
```

ntarocco commented:
Pushed a new commit to see diff, please squash when merging:
- fixed children bug, added tests, improve readability
- upgraded rollup
- added browser compat in package.json, which fixed the babel transpiling to ES6

ThiefMaster commented on src/overridable.js:
```diff
@@ -31,23 +31,27 @@ export function parametrize(Component, extraProps) {
-    child
-    ? React.cloneElement(child, childProps)
-    : null;
+  const hasOverridenCmp = id in overriddenCmps;
```
If you go that way I think you should use `overriddenCmps.hasOwnProperty(id)` :p

Btw, what about `Component` instead of `Cmp`? The latter sounds more like Comparison than Component to me...

ThiefMaster replied:
Actually, `hasOverridenCmp` is used only once so I'd just move the check in the line below...
